### DISCUSSION TITLE
Accessing bodyInterface inside a contact listener causes a crash.

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#160] [General]
I'm working on conveyor belt systems from the example and already had my own method to get rotation which uses `BodyInterface.GetRotation(this.BodyID)`

I couldn't figure out what was crashing things until I isolated that. 

When you call `Body.GetRotation()` it works fine, but if you try and access `BodyInterface` it crashes. 

I confirmed this behavior with `AddImpulse()` as well. 

Edited to add:
The crash is caused when you are inside:
```ts
contactListener.OnContactAdded = (body1, body2, manifold, settings) => {
// accessing bodyInterface here will crash things
}

````

**Comment by jrouwe** [comment]
If you use BodyInterface within a ContactListener callback you'll do a double-lock of the body because the physics sim has locked all bodies at that point. This is documented [here](https://jrouwe.github.io/JoltPhysics/class_contact_listener.html#aeb1d7908c8556ce24448ac572d7ad944).

**Reply by DennisSmolek** [reply]
So for a teleporter I was using `SetPosition()` to move the object. But that only exists on the BodyInterface. 

Should I just use `MoveKinematic()` with a near 0 time? Would that move through other bodies or would it get stuck?

I can wire up a "should teleport" thing that gets run every loop, but there may be a better method?

**Reply by LeXXik** [reply]
Just don't do it inside the callback. You can pool the actions needed to be done on callback and execute them after the physics update is done. You should not modify bodies inside the callbacks, treat them as read-only.
